[Site Isolation] Make setPageScaleFactor async

nmahendru opened the pull request:
#### 3b2b06f92fc7a57e191fa92e8e7e826be40319de
<pre>
[Site Isolation] Make setPageScaleFactor async
<a href="https://bugs.webkit.org/show_bug.cgi?id=277046">https://bugs.webkit.org/show_bug.cgi?id=277046</a>
<a href="https://rdar.apple.com/132440505">rdar://132440505</a>

Reviewed by Alex Christensen and Charlie Wolfe.

Also move it to TestRunner as that seems a better suited place for it. All Layout tests that use this API have been refactored.
I have used whatever strategy possible to not have diffs in the -expected files as much as possible.
That is why some tests are using waitUntilDone and some tests are using jsTestIsAsync.

* LayoutTests/accessibility/tree-update-with-dirty-layout.html:
* LayoutTests/compositing/canvas/hidpi-canvas-backing-store-invalidation-2.html:
* LayoutTests/compositing/geometry/fixed-position-composited-page-scale-down.html:
* LayoutTests/compositing/geometry/fixed-position-composited-page-scale-scroll.html:
* LayoutTests/compositing/geometry/fixed-position-composited-page-scale-smaller-than-viewport-expected.html:
* LayoutTests/compositing/geometry/fixed-position-composited-page-scale-smaller-than-viewport.html:
* LayoutTests/compositing/geometry/fixed-position-composited-page-scale.html:
* LayoutTests/compositing/geometry/fixed-position-iframe-composited-page-scale-down.html:
* LayoutTests/compositing/geometry/fixed-position-iframe-composited-page-scale.html:
* LayoutTests/compositing/geometry/fixed-position-transform-composited-page-scale-down.html:
* LayoutTests/compositing/geometry/fixed-position-transform-composited-page-scale.html:
* LayoutTests/compositing/layer-creation/fixed-position-out-of-view-scaled-iframe-scroll.html:
* LayoutTests/compositing/layer-creation/fixed-position-out-of-view-scaled-iframe.html:
* LayoutTests/compositing/layer-creation/fixed-position-out-of-view-scaled-scroll.html:
* LayoutTests/compositing/layer-creation/fixed-position-out-of-view-scaled.html:
* LayoutTests/compositing/overflow/overflow-scaled-descendant-overlapping.html:
* LayoutTests/compositing/repaint/page-scale-repaint.html:
* LayoutTests/compositing/scaling/tiled-layer-recursion.html:
* LayoutTests/css3/filters/backdrop-filter-page-scale-expected.html:
* LayoutTests/css3/filters/backdrop-filter-page-scale.html:
* LayoutTests/fast/css3-text/css3-text-decoration/text-decoration-scaled-expected.html:
* LayoutTests/fast/css3-text/css3-text-decoration/text-decoration-scaled.html:
* LayoutTests/fast/dom/Element/scale-page-bounding-client-rect-in-frame.html:
* LayoutTests/fast/dom/Element/scale-page-bounding-client-rect.html:
* LayoutTests/fast/dom/Element/scale-page-client-rects-in-frame.html:
* LayoutTests/fast/dom/Element/scale-page-client-rects.html:
* LayoutTests/fast/dom/Range/scale-page-bounding-client-rect.html:
* LayoutTests/fast/dom/Range/scale-page-client-rects.html:
* LayoutTests/fast/dom/elementFromPoint-scaled-scrolled-layout-viewport-expected.txt:
* LayoutTests/fast/dom/elementFromPoint-scaled-scrolled-layout-viewport.html:
* LayoutTests/fast/dom/elementFromPoint-scaled-scrolled.html:
* LayoutTests/fast/dom/iframe-inner-size-scaling.html:
* LayoutTests/fast/dom/window-inner-size-scaling-expected.txt:
* LayoutTests/fast/dom/window-inner-size-scaling.html:
* LayoutTests/fast/dom/window-scroll-scaling-expected.txt:
* LayoutTests/fast/dom/window-scroll-scaling.html:
* LayoutTests/fast/events/autoscroll-when-zoomed.html:
* LayoutTests/fast/events/drag-select-when-zoomed-with-header.html:
* LayoutTests/fast/events/drag-select-when-zoomed.html:
* LayoutTests/fast/events/page-scaled-mouse-click-iframe.html:
* LayoutTests/fast/events/page-scaled-mouse-click.html:
* LayoutTests/fast/events/scale-and-scroll-body.html:
* LayoutTests/fast/events/scale-and-scroll-iframe-body.html:
* LayoutTests/fast/events/scale-and-scroll-iframe-window.html:
* LayoutTests/fast/events/scale-and-scroll-window.html:
* LayoutTests/fast/events/scroll-in-scaled-page-with-overflow-hidden.html:
* LayoutTests/fast/events/touch/gesture/touch-gesture-scroll-div-scaled.html:
* LayoutTests/fast/events/touch/page-scaled-touch-gesture-click.html:
* LayoutTests/fast/events/touch/touch-scaled-scrolled.html:
* LayoutTests/fast/frames/frame-set-rotation-hit.html:
* LayoutTests/fast/frames/frame-set-scaling-hit.html:
* LayoutTests/fast/frames/iframe-double-scale-contents.html:
* LayoutTests/fast/repaint/background-scaling.html:
* LayoutTests/fast/repaint/scale-page-shrink.html:
* LayoutTests/fast/scrolling/mac/async-scroll-overflow-rtl-zoomed.html:
* LayoutTests/fast/scrolling/mac/overflow-zoomed-document.html:
* LayoutTests/fast/scrolling/scroll-to-anchor-zoomed-header.html:
* LayoutTests/fast/text/descent-clip-in-scaled-page-expected.html:
* LayoutTests/fast/text/descent-clip-in-scaled-page.html:
* LayoutTests/fast/transforms/selection-bounds-in-transformed-view.html:
* LayoutTests/imported/blink/fast/backgrounds/root-background-with-page-scaled-out-expected.html:
* LayoutTests/imported/blink/fast/backgrounds/root-background-with-page-scaled-out.html:
* LayoutTests/scrollingcoordinator/resources/non-fast-scrollable-region-testing.js:
(async runNonFastScrollableRegionTest):
(runNonFastScrollableRegionTest): Deleted.
* LayoutTests/svg/as-background-image/tiled-background-image-expected.html:
* LayoutTests/svg/as-background-image/tiled-background-image.html:
* LayoutTests/svg/as-image/image-respects-pageScaleFactor-change.html:
* LayoutTests/svg/as-image/image-respects-pageScaleFactor.html:
* LayoutTests/tiled-drawing/fixed-layout-size-fixed-attachment-cover.html:
* LayoutTests/tiled-drawing/fixed-layout-size-fixed-attachment-local.html:
* LayoutTests/tiled-drawing/header-and-footer-hit-testing-with-page-scale.html:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setPageScaleFactor): Deleted.
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageSetPageScaleFactor):
* Source/WebKit/UIProcess/API/C/WKPagePrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setPageScaleFactor):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setPageScaleFactor):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
(WebChromeClient::requestCookieConsent):
* Tools/DumpRenderTree/TestRunner.cpp:
(setPageScaleFactorCallback):
(TestRunner::staticFunctions):
* Tools/DumpRenderTree/TestRunner.h:
* Tools/DumpRenderTree/mac/TestRunnerMac.mm:
(TestRunner::setPageScaleFactor):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::setPageScaleFactor):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::didReceiveAsyncMessageFromInjectedBundle):
(WTR::TestController::setPageScaleFactor):
* Tools/WebKitTestRunner/TestController.h:

Canonical link: <a href="https://commits.webkit.org/282170@main">https://commits.webkit.org/282170@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2dc6d74e77fe18113019d5c1548ada6d95edf662

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62362 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41717 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/55/builds/14957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/66346 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/12911 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/64482 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/49403 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13247 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/66346 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/12911 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65431 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/49403 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/55/builds/14957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/66346 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/49403 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/55/builds/14957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/11842 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/49403 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/14957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/68072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/87/builds/6305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/11358 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/68072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/86/builds/6333 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/55/builds/14957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/68072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9380 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/37517 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/38602 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/39699 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/38346 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
```diff
@@ -541,10 +541,13 @@ class WebChromeClient final : public WebCore::ChromeClient {
     mutable bool m_cachedMainFrameHasVerticalScrollbar { false };
 
     WeakRef<WebPage> m_page;
+
```
unnecessary change.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -540,6 +540,7 @@ class WebPage : public API::ObjectImpl<API::Object::Type::BundlePage>, public IP
 #endif
 
     void simulateClickOverFirstMatchingTextInViewportWithUserInteraction(const String& targetText, CompletionHandler<void(bool)>&&);
+    void setPageScaleFactor(float, int, int, CompletionHandler<void()>&&);
```
We should not omit the names of the 3 first parameters as they are not obvious.

cdumez commented on Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
```diff
@@ -1173,3 +1173,11 @@ - (void)setIsSelected:(BOOL)isSelected;
+
+void WebChromeClient::setPageScaleFactor(float scaleFactor, int x, int y, CompletionHandler<void()>&& completionHandler)
+{
+    if (auto page = [m_webView page])
```
`auto` -> `RefPtr`

cdumez commented on Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.h:
```diff
@@ -258,5 +258,7 @@ class WebChromeClient : public WebCore::ChromeClient {
     bool m_mockVideoPresentationModeEnabled { false };
 #endif
 
+    void setPageScaleFactor(float, int, int, CompletionHandler<void()>&&) final;
```
We should not omit the name of the first 3 parameters.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
```diff
@@ -819,4 +819,6 @@ GenerateSyntheticEditingCommand(enum:uint8_t WebKit::SyntheticEditingCommandType
 
     SimulateClickOverFirstMatchingTextInViewportWithUserInteraction(String targetText) -> (bool success)
+
+    SetPageScaleFactor(float scaleFactor, int x, int y) -> ()
```
Why aren't we sending an `IntPoint` over IPC instead of `x` & `y` separately? Seems that what the Page needs eventually anyway.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.messages.in:
```diff
@@ -647,4 +647,6 @@ CreateAppHighlightInSelectedRange)
 
     SetAllowsLayoutViewportHeightExpansion(bool newValue)
+
+    SetPageScaleFactor(float scaleFactor, int x, int y) -> ()
```
We should consider passing a `WebCore::IntPoint origin` instead of x / y

cdumez replied:
Seems unfortunate to be added new WebPageProxy IPC just for testing. I think the IPC should be in the TestRunner instead and the API should be in TestRunner.idl instead.

cdumez commented on Source/WebCore/testing/Internals.idl:
```diff
@@ -815,7 +815,7 @@ enum RenderingMode {
     DOMString pageSizeAndMarginsInPixels(long pageIndex, long width, long height, long marginTop, long marginRight, long marginBottom, long marginLeft);
 
-    undefined setPageScaleFactor(unrestricted float scaleFactor, long x, long y);
+    Promise<undefined> setPageScaleFactor(unrestricted float scaleFactor, long x, long y);
```
Have we considered moving this to TestRunner.idl? Going to the UIProcess is exactly what testRunner API does. Whereas internals.idl API is usually expected to set some state directly inside the WebProcess.

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
```diff
@@ -824,4 +824,6 @@ GenerateSyntheticEditingCommand(enum:uint8_t WebKit::SyntheticEditingCommandType
 
     SimulateClickOverFirstMatchingTextInViewportWithUserInteraction(String targetText) -> (bool success)
+
+    SetPageScaleFactor(float scaleFactor, WebCore::IntPoint point) -> ()
```
We don’t need to introduce a new test-only IPC message for this. You could replace `Messages::WebPageProxy::PageScaleFactorDidChange` in `WebPage::scalePage` with a completion handler and use the existing IPC message (`Messages::WebPage::ScalePage`).

nmahendru replied:
As discussed we can't do this as:
1. it's not a 1-1 replacement for the setPageScaleFactor in Internals.(scalePage is doing more).
2. All callers of scalePage don't really have a completionHandler and there are a bunch of them and rooting them out is probably good for another PR. Scope TBD

charliewolfe commented on Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
```diff
@@ -463,4 +463,5 @@ interface TestRunner {
     Promise<undefined> flushConsoleLogs();
 
     Promise<undefined> setTopContentInset(double contentInset);
+    Promise<undefined> setPageScaleFactor(unrestricted float scaleFactor, long x, long y);
```
Change the `scaleFactor` type to `double` to match whats in C++.

charliewolfe commented on Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
```diff
@@ -2121,6 +2121,16 @@ void TestRunner::flushConsoleLogs(JSContextRef context, JSValueRef callback)
+        { "x", toWK(x) },
+        { "y", toWK(y) },
+        }),
+        callback);
```
```suggestion
    }), callback);
```

achristensen07 commented on Source/WebKit/UIProcess/API/C/WKPagePrivate.h:
```diff
@@ -218,7 +218,8 @@ WK_EXPORT void WKPageSetPermissionLevelForTesting(WKPageRef page, WKStringRef or
 WK_EXPORT void WKPageSetTopContentInsetForTesting(WKPageRef page, float contentInset, void* context, WKPageSetTopContentInsetForTestingFunction callback);
-
+typedef void (*WKPageSetPageScaleFactorFunction)(void* functionContext);
+WK_EXPORT void WKPageSetPageScaleFactor(WKPageRef page, float scaleFactor, int x, int y, void* context, WKPageSetPageScaleFactorFunction completionHandler);
```
Let's add ForTesting.

charliewolfe replied:
Other test-only C SPI has the "ForTesting" suffix

achristensen07 commented on Source/WebKit/UIProcess/API/C/WKPage.cpp:
```diff
@@ -3305,3 +3305,10 @@ void WKPageSetTopContentInsetForTesting(WKPageRef pageRef, float contentInset, v
+
+void WKPageSetPageScaleFactor(WKPageRef pageRef, float scaleFactor, int x, int y, void* context, WKPageSetPageScaleFactorFunction completionHandler)
+{
+    toImpl(pageRef)->setPageScaleFactor(scaleFactor, WebCore::IntPoint { x, y }, [context, completionHandler] {
```
Can we use pageForTesting like WKPageSetTopContentInsetForTesting?

achristensen07 commented on Tools/DumpRenderTree/TestRunner.cpp:
```diff
@@ -1844,6 +1844,22 @@ static JSValueRef runUIScriptCallback(JSContextRef context, JSObjectRef, JSObjec
+        ASSERT(!*exception);
+        int y = JSValueToInt64(context, arguments[2], exception);
+        ASSERT(!*exception);
+        // TestRunnerMac.mm
```
Let's remove this comment.

charliewolfe commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2488,6 +2488,7 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
     void closeCurrentTypingCommand();
 
     void simulateClickOverFirstMatchingTextInViewportWithUserInteraction(String&& targetText, CompletionHandler<void(bool)>&&);
+    void setPageScaleFactor(float scaleFactor, WebCore::IntPoint, CompletionHandler<void()>&&);
```
You could move this to `WebPageProxyTesting`

achristensen07 commented on LayoutTests/fast/events/scale-and-scroll-body.html:
```diff
@@ -27,19 +29,22 @@
-              window.internals.setPageScaleFactor(1, 30, 30);
+              if (window.testRunner)
+                  await window.testRunner.setPageScaleFactor(1, 30, 30);
               shouldBe("window.document.scrollingElement.scrollTop", "30");
```
This test needs investigation.  It failed in stress EWS and I don't see why.

achristensen07 commented on LayoutTests/fast/text/descent-clip-in-scaled-page.html:
```diff
@@ -21,23 +21,29 @@
 </style>
 <script>
-    if (window.testRunner && window.testRunner.setTextSubpixelPositioning)
+    if (window.testRunner && window.testRunner.setTextSubpixelPositioning){
```
Ditto.

charliewolfe commented on Source/WebKit/UIProcess/API/C/WKPagePrivate.h:
```diff
@@ -218,7 +218,8 @@ WK_EXPORT void WKPageSetPermissionLevelForTesting(WKPageRef page, WKStringRef or
 WK_EXPORT void WKPageSetTopContentInsetForTesting(WKPageRef page, float contentInset, void* context, WKPageSetTopContentInsetForTestingFunction callback);
-
+typedef void (*WKPageSetPageScaleFactorForTestingFunction)(void* functionContext);
+WK_EXPORT void WKPageSetPageScaleFactorForTesting(WKPageRef page, float scaleFactor, int x, int y, void* context, WKPageSetPageScaleFactorForTestingFunction completionHandler);
```
This should take a `WKPoint` instead of `int x, int y`, similar to `WKPageSetScaleFactor`.

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -713,6 +713,7 @@ class WebPage : public API::ObjectImpl<API::Object::Type::BundlePage>, public IP
 
     void screenPropertiesDidChange();
 
+    void setPageScaleFactor(double scale, const WebCore::IntPoint& origin, CompletionHandler<void()>&&);
```
This function is not used, remove.

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
```diff
@@ -828,4 +827,5 @@ GenerateSyntheticEditingCommand(enum:uint8_t WebKit::SyntheticEditingCommandType
     CloseCurrentTypingCommand()
 
     SimulateClickOverFirstMatchingTextInViewportWithUserInteraction(String targetText) -> (bool success)
+
```
Undo the changes in this file

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebPageTesting.cpp:
```diff
@@ -113,6 +113,16 @@ void WebPageTesting::setTopContentInset(float contentInset, CompletionHandler<vo
+{
+    RefPtr page = m_page->corePage();
+    if (!page)
+        return;
```
```suggestion
        return completionHandler();
```

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebPageTesting.cpp:
```diff
@@ -113,6 +113,16 @@ void WebPageTesting::setTopContentInset(float contentInset, CompletionHandler<vo
     completionHandler();
 }
 
+void WebPageTesting::setPageScaleFactor(double scale, const IntPoint& origin, CompletionHandler<void()>&& completionHandler)
```
`IntPoint` is 8 bytes, so just pass by value.

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebPageTesting.cpp:
```diff
@@ -113,6 +113,16 @@ void WebPageTesting::setTopContentInset(float contentInset, CompletionHandler<vo
+    completionHandler();
+}
+
+
```
Remove extra new line

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebPageTesting.h:
```diff
@@ -26,6 +26,7 @@
 #pragma once
 
 #include "MessageReceiver.h"
+#include <WebCore/IntPoint.h>
```
Remove this and forward declare `IntPoint`

charliewolfe commented on Tools/WebKitTestRunner/TestController.cpp:
```diff
@@ -1988,6 +1988,15 @@ void TestController::didReceiveAsyncMessageFromInjectedBundle(WKStringRef messag
         return completionHandler(nullptr);
 
+    if (WKStringIsEqualToUTF8CString(messageName, "SetPageScaleFactor")) {
+
```
Remove extra new line

charliewolfe reviewed (commented):
Looks good, but `compositing/geometry/fixed-position-composited-page-scale-smaller-than-viewport.html` is still failing on wk1.

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
Undo changes in this file

charliewolfe commented on Source/WebKit/UIProcess/WebPageProxyTesting.h:
```diff
@@ -27,6 +27,7 @@
 
 #include "MessageSender.h"
 #include <WebCore/FrameIdentifier.h>
+#include <WebCore/IntPoint.h>
```
Just forward declare `IntPoint`

charliewolfe commented on Tools/DumpRenderTree/mac/TestRunnerMac.mm:
```diff
@@ -474,6 +475,20 @@ static inline size_t indexOfSeparatorAfterDirectoryName(const std::string& direc
 
+JSValueRef TestRunner::alwaysResolvePromise(JSContextRef context)
+{
+    JSContext * jsContext = [JSContext contextWithJSGlobalContextRef:toGlobalRef(toJS(context))];
```
```suggestion
    JSContext *jsContext = [JSContext contextWithJSGlobalContextRef:toGlobalRef(toJS(context))];
```

charliewolfe commented on Tools/WebKitTestRunner/TestController.cpp:
```diff
@@ -3489,6 +3497,11 @@ void TestController::clearAppPrivacyReportTestingData()
 
+void TestController::setPageScaleFactor(float scaleFactor, int x, int y, CompletionHandler<void(WKTypeRef)>&& completionHandler)
+{
+    WKPageSetPageScaleFactorForTesting(mainWebView()->page(), scaleFactor, WKPoint { static_cast<double>(x), static_cast<double>(y) }, completionHandler.leak(), adoptAndCallCompletionHandler);
```
WKPoint constructor takes a float, not a double. Just remove the `static_cast`.
```suggestion
    WKPageSetPageScaleFactorForTesting(mainWebView()->page(), scaleFactor, WKPoint { x, y }, completionHandler.leak(), adoptAndCallCompletionHandler);
```

nmahendru replied:
that is not true. see https://github.com/WebKit/WebKit/blob/8cf3d3cdb4d73ecfd90b0e44d60fec1053c2df48/Source/WebKit/Shared/API/c/WKGeometry.h#L35
It takes a double.

achristensen07 replied:
int can be implicitly converted to a double in this case.

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebPageTesting.h:
```diff
@@ -59,6 +63,8 @@ class WebPageTesting : public IPC::MessageReceiver {
 
     void setTopContentInset(float, CompletionHandler<void()>&&);
 
+    void setPageScaleFactor(double scale, const WebCore::IntPoint origin, CompletionHandler<void()>&&);
```
origin doesn't need const, since we are passing by value

charliewolfe reviewed (approved):
r=me assuming EWS is green

achristensen07 commented on Source/WebKit/UIProcess/API/C/WKPage.cpp:
```diff
@@ -3287,3 +3287,14 @@ void WKPageSetTopContentInsetForTesting(WKPageRef pageRef, float contentInset, v
+void WKPageSetPageScaleFactorForTesting(WKPageRef pageRef, float scaleFactor, WKPoint point, void* context, WKPageSetPageScaleFactorForTestingFunction completionHandler)
+{
+    auto* pageForTesting = toImpl(pageRef)->pageForTesting();
+    if (!pageForTesting)
```
I'm not sure if we want this case since it's for testing and this should never happen.  Fine either way.

nmahendru committed:
[Site Isolation] Make setPageScaleFactor async
https://bugs.webkit.org/show_bug.cgi?id=277046
rdar://132440505

Reviewed by Alex Christensen and Charlie Wolfe.

Also move it to TestRunner as that seems a better suited place for it. All Layout tests that use this API have been refactored.
I have used whatever strategy possible to not have diffs in the -expected files as much as possible.
That is why some tests are using waitUntilDone and some tests are using jsTestIsAsync.

* LayoutTests/accessibility/tree-update-with-dirty-layout.html:
* LayoutTests/compositing/canvas/hidpi-canvas-backing-store-invalidation-2.html:
* LayoutTests/compositing/geometry/fixed-position-composited-page-scale-down.html:
* LayoutTests/compositing/geometry/fixed-position-composited-page-scale-scroll.html:
* LayoutTests/compositing/geometry/fixed-position-composited-page-scale-smaller-than-viewport-expected.html:
* LayoutTests/compositing/geometry/fixed-position-composited-page-scale-smaller-than-viewport.html:
* LayoutTests/compositing/geometry/fixed-position-composited-page-scale.html:
* LayoutTests/compositing/geometry/fixed-position-iframe-composited-page-scale-down.html:
* LayoutTests/compositing/geometry/fixed-position-iframe-composited-page-scale.html:
* LayoutTests/compositing/geometry/fixed-position-transform-composited-page-scale-down.html:
* LayoutTests/compositing/geometry/fixed-position-transform-composited-page-scale.html:
* LayoutTests/compositing/layer-creation/fixed-position-out-of-view-scaled-iframe-scroll.html:
* LayoutTests/compositing/layer-creation/fixed-position-out-of-view-scaled-iframe.html:
* LayoutTests/compositing/layer-creation/fixed-position-out-of-view-scaled-scroll.html:
* LayoutTests/compositing/layer-creation/fixed-position-out-of-view-scaled.html:
* LayoutTests/compositing/overflow/overflow-scaled-descendant-overlapping.html:
* LayoutTests/compositing/repaint/page-scale-repaint.html:
* LayoutTests/compositing/scaling/tiled-layer-recursion.html:
* LayoutTests/css3/filters/backdrop-filter-page-scale-expected.html:
* LayoutTests/css3/filters/backdrop-filter-page-scale.html:
* LayoutTests/fast/css3-text/css3-text-decoration/text-decoration-scaled-expected.html:
* LayoutTests/fast/css3-text/css3-text-decoration/text-decoration-scaled.html:
* LayoutTests/fast/dom/Element/scale-page-bounding-client-rect-in-frame.html:
* LayoutTests/fast/dom/Element/scale-page-bounding-client-rect.html:
* LayoutTests/fast/dom/Element/scale-page-client-rects-in-frame.html:
* LayoutTests/fast/dom/Element/scale-page-client-rects.html:
* LayoutTests/fast/dom/Range/scale-page-bounding-client-rect.html:
* LayoutTests/fast/dom/Range/scale-page-client-rects.html:
* LayoutTests/fast/dom/elementFromPoint-scaled-scrolled-layout-viewport-expected.txt:
* LayoutTests/fast/dom/elementFromPoint-scaled-scrolled-layout-viewport.html:
* LayoutTests/fast/dom/elementFromPoint-scaled-scrolled.html:
* LayoutTests/fast/dom/iframe-inner-size-scaling.html:
* LayoutTests/fast/dom/window-inner-size-scaling-expected.txt:
* LayoutTests/fast/dom/window-inner-size-scaling.html:
* LayoutTests/fast/dom/window-scroll-scaling-expected.txt:
* LayoutTests/fast/dom/window-scroll-scaling.html:
* LayoutTests/fast/events/autoscroll-when-zoomed.html:
* LayoutTests/fast/events/drag-select-when-zoomed-with-header.html:
* LayoutTests/fast/events/drag-select-when-zoomed.html:
* LayoutTests/fast/events/page-scaled-mouse-click-iframe.html:
* LayoutTests/fast/events/page-scaled-mouse-click.html:
* LayoutTests/fast/events/scale-and-scroll-body.html:
* LayoutTests/fast/events/scale-and-scroll-iframe-body.html:
* LayoutTests/fast/events/scale-and-scroll-iframe-window.html:
* LayoutTests/fast/events/scale-and-scroll-window.html:
* LayoutTests/fast/events/scroll-in-scaled-page-with-overflow-hidden.html:
* LayoutTests/fast/events/touch/gesture/touch-gesture-scroll-div-scaled.html:
* LayoutTests/fast/events/touch/page-scaled-touch-gesture-click.html:
* LayoutTests/fast/events/touch/touch-scaled-scrolled.html:
* LayoutTests/fast/frames/frame-set-rotation-hit.html:
* LayoutTests/fast/frames/frame-set-scaling-hit.html:
* LayoutTests/fast/frames/iframe-double-scale-contents.html:
* LayoutTests/fast/repaint/background-scaling.html:
* LayoutTests/fast/repaint/scale-page-shrink.html:
* LayoutTests/fast/scrolling/mac/async-scroll-overflow-rtl-zoomed.html:
* LayoutTests/fast/scrolling/mac/overflow-zoomed-document.html:
* LayoutTests/fast/scrolling/scroll-to-anchor-zoomed-header.html:
* LayoutTests/fast/text/descent-clip-in-scaled-page-expected.html:
* LayoutTests/fast/text/descent-clip-in-scaled-page.html:
* LayoutTests/fast/transforms/selection-bounds-in-transformed-view.html:
* LayoutTests/imported/blink/fast/backgrounds/root-background-with-page-scaled-out-expected.html:
* LayoutTests/imported/blink/fast/backgrounds/root-background-with-page-scaled-out.html:
* LayoutTests/scrollingcoordinator/resources/non-fast-scrollable-region-testing.js:
(async runNonFastScrollableRegionTest):
(runNonFastScrollableRegionTest): Deleted.
* LayoutTests/svg/as-background-image/tiled-background-image-expected.html:
* LayoutTests/svg/as-background-image/tiled-background-image.html:
* LayoutTests/svg/as-image/image-respects-pageScaleFactor-change.html:
* LayoutTests/svg/as-image/image-respects-pageScaleFactor.html:
* LayoutTests/tiled-drawing/fixed-layout-size-fixed-attachment-cover.html:
* LayoutTests/tiled-drawing/fixed-layout-size-fixed-attachment-local.html:
* LayoutTests/tiled-drawing/header-and-footer-hit-testing-with-page-scale.html:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setPageScaleFactor): Deleted.
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageSetPageScaleFactor):
* Source/WebKit/UIProcess/API/C/WKPagePrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setPageScaleFactor):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setPageScaleFactor):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
(WebChromeClient::requestCookieConsent):
* Tools/DumpRenderTree/TestRunner.cpp:
(setPageScaleFactorCallback):
(TestRunner::staticFunctions):
* Tools/DumpRenderTree/TestRunner.h:
* Tools/DumpRenderTree/mac/TestRunnerMac.mm:
(TestRunner::setPageScaleFactor):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::setPageScaleFactor):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::didReceiveAsyncMessageFromInjectedBundle):
(WTR::TestController::setPageScaleFactor):
* Tools/WebKitTestRunner/TestController.h:

Canonical link: https://commits.webkit.org/282170@main
